File: frontend/src/lib/api/challenges.ts
import { getApiBaseUrl } from "./config";

export type VisibleTestResult = {
  id: string;
  input: unknown;
  expected: unknown;
  actual: unknown;
  passed: boolean;
  failure_reason: string | null;
};

export type RunChallengeResponse = {
  challenge: string;
  challenge_id: string;
  version: string;
  passed: number;
  total: number;
  accuracy: number;
  tests: VisibleTestResult[];
};

export type SubmitChallengeResponse = {
  challenge: string;
  version: string;
  passed: number;
  total: number;
  accuracy: number;
  prompt_tokens: number;
  efficiency: number;
  score: number;
  stars: number;
  xp_earned: number;
  total_xp: number;
  best_score: number;
  best_stars: number;
  completed: boolean;
};

export type CurrentUserResponse = {
  id: string;
  email: string | null;
  username: string | null;
};

export type ChallengeProgress = {
  challenge: string;
  best_score: number;
  best_stars: number;
  attempts: number;
  completed: boolean;
  completed_at: string | null;
};

export type CurrentUserProgressResponse = {
  total_xp: number;
  challenges_completed: number;
  stars_earned: number;
  challenges: ChallengeProgress[];
};

export type ChallengeStatus = "locked" | "available" | "completed" | "mastered";

export type ChallengeListItem = {
  id: string;
  slug: string;
  title: string;
  track: string;
  difficulty: string;
  order: number;
  status: ChallengeStatus;
  best_score: number | null;
  best_stars: number;
  attempts: number;
  completed: boolean;
};

export type ChallengeListResponse = { challenges: ChallengeListItem[] };

export type ChallengeDetailResponse = ChallengeListItem & {
  description: string;
  objective: string;
  constraints: string[];
  version: string;
  prompt_token_limit: number | null;
  examples: Array<{ input: unknown; expected: unknown; explanation: string | null }>;
};

export class ApiError extends Error {
  constructor(
    message: string,
    readonly status: number,
    readonly code: string | null = null,
  ) {
    super(message);
    this.name = "ApiError";
  }
}

export class AuthRequiredError extends Error {
  constructor() {
    super("Sign in before submitting a challenge.");
    this.name = "AuthRequiredError";
  }
}

async function parseError(response: Response): Promise<ApiError> {
  try {
    const body: unknown = await response.json();
    if (typeof body === "object" && body !== null && "error" in body) {
      const error = body.error;
      if (typeof error === "object" && error !== null) {
        const message = "message" in error && typeof error.message === "string"
          ? error.message
          : "The request failed.";
        const code = "code" in error && typeof error.code === "string" ? error.code : null;
        return new ApiError(message, response.status, code);
      }
    }
  } catch {
    // Fall through to a safe generic error.
  }
  return new ApiError("The request failed.", response.status);
}

async function postPrompt<T>(
  path: string,
  prompt: string,
  accessToken: string | null,
  fetcher: typeof fetch,
): Promise<T> {
  const headers: Record<string, string> = { "Content-Type": "application/json" };
  if (accessToken !== null) {
    headers.Authorization = `Bearer ${accessToken}`;
  }
  let response: Response;
  try {
    response = await fetcher(`${getApiBaseUrl()}${path}`, {
      method: "POST",
      headers,
      body: JSON.stringify({ prompt }),
    });
  } catch {
    throw new ApiError("The backend is unavailable.", 0);
  }
  if (!response.ok) {
    throw await parseError(response);
  }
  return (await response.json()) as T;
}

export function runChallenge(
  challengeSlug: string,
  prompt: string,
  fetcher: typeof fetch = fetch,
): Promise<RunChallengeResponse> {
  return postPrompt(
    `/api/v1/challenges/${encodeURIComponent(challengeSlug)}/run`,
    prompt,
    null,
    fetcher,
  );
}

export function submitChallenge(
  challengeSlug: string,
  prompt: string,
  accessToken: string | null,
  fetcher: typeof fetch = fetch,
): Promise<SubmitChallengeResponse> {
  if (accessToken === null) {
    throw new AuthRequiredError();
  }
  return postPrompt(
    `/api/v1/challenges/${encodeURIComponent(challengeSlug)}/submit`,
    prompt,
    accessToken,
    fetcher,
  );
}

async function getChallengeResource<T>(
  path: string,
  accessToken: string | null,
  fetcher: typeof fetch,
): Promise<T> {
  const headers: Record<string, string> = {};
  if (accessToken !== null) headers.Authorization = `Bearer ${accessToken}`;
  let response: Response;
  try {
    response = await fetcher(`${getApiBaseUrl()}${path}`, { headers, cache: "no-store" });
  } catch {
    throw new ApiError("The backend is unavailable.", 0);
  }
  if (!response.ok) throw await parseError(response);
  return (await response.json()) as T;
}

export function getChallenges(
  accessToken: string | null,
  fetcher: typeof fetch = fetch,
): Promise<ChallengeListResponse> {
  return getChallengeResource("/api/v1/challenges", accessToken, fetcher);
}

export function getChallengeDetail(
  challengeSlug: string,
  accessToken: string | null,
  fetcher: typeof fetch = fetch,
): Promise<ChallengeDetailResponse> {
  return getChallengeResource(
    `/api/v1/challenges/${encodeURIComponent(challengeSlug)}`,
    accessToken,
    fetcher,
  );
}

export async function getCurrentUser(
  accessToken: string,
  fetcher: typeof fetch = fetch,
): Promise<CurrentUserResponse> {
  let response: Response;
  try {
    response = await fetcher(`${getApiBaseUrl()}/api/v1/me`, {
      headers: { Authorization: `Bearer ${accessToken}` },
      cache: "no-store",
    });
  } catch {
    throw new ApiError("The backend is unavailable.", 0);
  }
  if (!response.ok) {
    throw await parseError(response);
  }
  return (await response.json()) as CurrentUserResponse;
}

export async function getCurrentUserProgress(
  accessToken: string,
  fetcher: typeof fetch = fetch,
): Promise<CurrentUserProgressResponse> {
  let response: Response;
  try {
    response = await fetcher(`${getApiBaseUrl()}/api/v1/me/progress`, {
      headers: { Authorization: `Bearer ${accessToken}` },
      cache: "no-store",
    });
  } catch {
    throw new ApiError("The backend is unavailable.", 0);
  }
  if (!response.ok) {
    throw await parseError(response);
  }
  return (await response.json()) as CurrentUserProgressResponse;
}

export function getReadableError(error: unknown): string {
  if (error instanceof AuthRequiredError || error instanceof ApiError) {
    return error.message;
  }
  return "The request failed unexpectedly.";
}

File: backend/app/api/schemas/challenges.py
from pydantic import BaseModel, ConfigDict, Field, JsonValue, field_validator

from app.application.challenges.models import RunChallengeResult, SubmitChallengeResult
from app.domains.challenges.models import PlayableChallenge
from app.domains.progression import ChallengeAccess


class ChallengePromptRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    prompt: str = Field(min_length=1)

    @field_validator("prompt")
    @classmethod
    def prompt_must_not_be_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("Prompt cannot be blank.")
        return value


class RunChallengeRequest(ChallengePromptRequest):
    pass


class SubmitChallengeRequest(ChallengePromptRequest):
    pass


class ChallengeListItemResponse(BaseModel):
    id: str
    slug: str
    title: str
    track: str
    difficulty: str
    order: int
    status: str
    best_score: float | None
    best_stars: int
    attempts: int
    completed: bool

    @classmethod
    def from_access(cls, access: ChallengeAccess) -> "ChallengeListItemResponse":
        playable = access.challenge
        return cls(
            id=playable.challenge.id,
            slug=playable.challenge.slug,
            title=playable.version.title,
            track=playable.challenge.track.value,
            difficulty=playable.version.difficulty.value,
            order=playable.challenge.order,
            status=access.status.value,
            best_score=(round(access.best_score, 2) if access.best_score is not None else None),
            best_stars=access.best_stars,
            attempts=access.attempts,
            completed=access.completed,
        )


class ChallengeListResponse(BaseModel):
    challenges: list[ChallengeListItemResponse]


class VisibleExampleResponse(BaseModel):
    input: JsonValue
    expected: JsonValue
    explanation: str | None


class ChallengeDetailResponse(BaseModel):
    id: str
    slug: str
    title: str
    description: str
    objective: str
    constraints: list[str]
    track: str
    difficulty: str
    order: int
    version: str
    prompt_token_limit: int | None
    status: str
    best_score: float | None
    best_stars: int
    attempts: int
    completed: bool
    examples: list[VisibleExampleResponse]

    @classmethod
    def from_domain(
        cls,
        playable: PlayableChallenge,
        access: ChallengeAccess,
    ) -> "ChallengeDetailResponse":
        return cls(
            id=playable.challenge.id,
            slug=playable.challenge.slug,
            title=playable.version.title,
            description=playable.version.description,
            objective=playable.version.objective,
            constraints=list(playable.version.constraints),
            track=playable.challenge.track.value,
            difficulty=playable.version.difficulty.value,
            order=playable.challenge.order,
            version=playable.version.version_id,
            prompt_token_limit=playable.version.prompt_token_limit,
            status=access.status.value,
            best_score=(round(access.best_score, 2) if access.best_score is not None else None),
            best_stars=access.best_stars,
            attempts=access.attempts,
            completed=access.completed,
            examples=[
                VisibleExampleResponse(
                    input=example.input,
                    expected=example.expected_output,
                    explanation=example.explanation,
                )
                for example in playable.version.visible_examples
            ],
        )


class VisibleTestRunResponse(BaseModel):
    id: str
    input: JsonValue
    expected: JsonValue
    actual: JsonValue
    passed: bool
    failure_reason: str | None


class RunChallengeResponse(BaseModel):
    challenge: str
    challenge_id: str
    version: str
    passed: int
    total: int
    accuracy: float
    tests: list[VisibleTestRunResponse]

    @classmethod
    def from_application_result(cls, result: RunChallengeResult) -> "RunChallengeResponse":
        return cls(
            challenge=result.challenge_slug,
            challenge_id=result.challenge_id,
            version=result.challenge_version_id,
            passed=result.passed_count,
            total=result.total_count,
            accuracy=round(result.accuracy, 2),
            tests=[
                VisibleTestRunResponse(
                    id=test.test_id,
                    input=test.input,
                    expected=test.expected_output,
                    actual=test.actual_output,
                    passed=test.passed,
                    failure_reason=(
                        test.failure_reason.value if test.failure_reason is not None else None
                    ),
                )
                for test in result.test_results
            ],
        )


class SubmitChallengeResponse(BaseModel):
    """Aggregate-only response; hidden per-test data has no representable field."""

    challenge: str
    version: str
    passed: int
    total: int
    accuracy: float
    prompt_tokens: int
    efficiency: float
    score: float
    stars: int
    xp_earned: int
    total_xp: int
    best_score: float
    best_stars: int
    completed: bool

    @classmethod
    def from_application_result(
        cls,
        result: SubmitChallengeResult,
    ) -> "SubmitChallengeResponse":
        return cls(
            challenge=result.challenge_slug,
            version=result.challenge_version_id,
            passed=result.passed_count,
            total=result.total_count,
            accuracy=round(result.accuracy, 2),
            prompt_tokens=result.prompt_tokens,
            efficiency=round(result.efficiency, 2),
            score=result.final_score,
            stars=result.stars,
            xp_earned=result.xp_earned,
            total_xp=result.total_xp,
            best_score=result.best_score,
            best_stars=result.best_stars,
            completed=result.completed,
        )
